rspace/src/main/scala/coop/rchain/rspace/trace/Event.scala

```diff
@@ -23,16 +25,18 @@ final case class COMM(
 ) extends Event
 
 object COMM {
-  def apply[C, A](
+  def apply[F[_]: Concurrent: ContextShift, C, A](
       dataCandidates: Seq[ConsumeCandidate[C, A]],
       consumeRef: Consume,
       peeks: SortedSet[Int],
-      produceCounters: (Seq[Produce]) => Map[Produce, Int]
-  ): COMM = {
+      produceCounters: (Seq[Produce]) => F[Map[Produce, Int]]
+  ): F[COMM] = {
     val produceRefs =
       dataCandidates.map(_.datum.source).sortBy(p => (p.channelsHash, p.hash, p.persistent))
 
-    COMM(consumeRef, produceRefs, peeks, produceCounters(produceRefs))
+    for {
+      counters <- produceCounters(produceRefs)
+    } yield COMM(consumeRef, produceRefs, peeks, counters)
```

Comment:
Now that function is returning value inside F context, it's executed differently.

When called immediately `produceRefs` is calculated.
```scala
val produceRefs =
  dataCandidates.map(_.datum.source).sortBy(p => (p.channelsHash, p.hash, p.persistent))
```
When F context is scheduled to execute (map, flatMap, ... sequence of terms) then `produceCounters` and resto of the code is executed. This means if F part of the function is never executed we have wasted `produceRefs` mapping and sorting.

To make `produceRefs` calculation _lazy_ and part of F context wrap it with `Sync[F].delay` which will take any `A` to `F[A]`.

So you need to start with
```scala
Sync[F]
  .delay(dataCandidates.map(_.datum.source).sortBy(p => (p.channelsHash, p.hash, p.persistent)))
```
and choose what syntax is more convenient `.map(...).flatMap(...)` or `for ... yield ...`. I usually try both to see which makes less nesting and splitting in multiple lines.